### PR TITLE
Add Expr.CustomFunction and datetime functions

### PR DESCRIPTION
**Pull request by jeffmay**
- Add Expr.CustomFunction and datetime functions
- Add TimeFunctions to ExprDsl
- Improve unit tests for Age from DateOfBirth
- Add ModifyTime typeclass for adding or subtracting from times
- Add CountTime typeclass for counting the difference between times
- Add now and today methods for injecting the current time from a clock
- Deprecate builder.embed and add better alternatives
- Improve CaptureP implicitNotFound message

### CODE REVIEW

File: core/src/main/scala/com/rallyhealth/vapors/core/dsl/TimeFunctions.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  ): Expr[V, T, P] = {
+    val argExpr = wrap(lhs, rhs).asHList.returnOutput
+    val fn: T :: D :: HNil => T = {
+      case lhs :: rhs :: HNil =>
+        ModifyTime.addTo(lhs, rhs)
+    }
+    Expr.CustomFunction(argExpr, "date_add", fn, captureResult)
+  }
+
+  // TODO: Can this be converted to Addition[T] if we make the second argument a different type? Maybe AddLeft[T, D]?
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/core/dsl/TimeFunctions.scala]: This is a pretty big change, I'm gonna make this change in a separate PR, targeted for 0.14.0.

File: core/src/main/scala/com/rallyhealth/vapors/core/time/CountTime.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      start: T,
+      end: T,
+      unit: TemporalUnit,
+    ): Long = start.until(end, unit)
+  }
+
+  @inline private def temporal[T <: Temporal]: CountTime[T, ChronoUnit] = new OfTemporal[T]
+
+  // TODO: Restrict invalid units by type
+  implicit val ofTemporal: CountTime[Temporal, ChronoUnit] = temporal
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/core/time/CountTime.scala:43]: There isn't a good type distinction between `ChronoUnit.SECONDS` and `ChronoUnit.YEARS` or `ChronoUnit.FOREVER`, but it will blow up at runtime if you attempt to count the amount of time in years or "forever"s (whatever that means) between `Instant`s. This would require creating my own units that map to other time units, but maintain compile-time safety. I will take a crack at this when I have more time.

File: core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    override def subtractFrom(
+      temporal: T,
+      amount: D,
+    ): T = amount.subtractFrom(temporal).asInstanceOf[T]
+  }
+
+  @inline private def temporal[T <: Temporal, D <: TemporalAmount]: ModifyTime[T, D] = new OfTemporal[T, D]
+
+  implicit val ofInstant: ModifyTime[Instant, Duration] = temporal
+  implicit val ofLocalDate: ModifyTime[LocalDate, Period] = temporal
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala:50]: This guarantees at compile-time that you do not attempt to modify unmodifiable fields. A `LocalDate` only supports adding or subtracting days, months, and years, which are the only fields that you can use to create a `Period`. Conversely, an `Instant` only supports adding or subtracting nanoseconds, milliseconds, seconds, minutes, hours, half-days, and days, which are the only fields you can use to create a `Duration`.

By limiting these two types to only allow modifying these more specific `TemporalAmount` subclasses, we can guarantee at compile-time that you can't add an amount that would fail at runtime.

File: core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  implicit val ofLocalDateTime: ModifyTime[LocalDateTime, TemporalAmount] = temporal
+  implicit val ofZonedDateTime: ModifyTime[ZonedDateTime, TemporalAmount] = temporal
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala]: These support any adding or subtracting any `TemporalAmount`, so we use contravariance to support this.

File: core/src/test/scala/com/rallyhealth/vapors/core/example/JoeSchmoe.scala [hunk 5]
```diff
@@ -9,7 +9,7 @@ object JoeSchmoe {
   val age = FactTypes.Age(32)
   val userRole = FactTypes.Role(Role.User)
   val adminRole = FactTypes.Role(Role.Admin)
-  val dateOfBirth = FactTypes.DateOfBirth(LocalDate.of(1987, 1, 1))
+  val dateOfBirth = FactTypes.DateOfBirth(LocalDate.of(1988, 8, 8))
```
Comment (jeffmay) [inline core/src/test/scala/com/rallyhealth/vapors/core/example/JoeSchmoe.scala:12]: This wasn't needed, but it seems easier to remember 8/8/88 and has a reasonable year in the past.

File: core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  private final class OfTemporal[T <: Temporal, -D <: TemporalAmount] extends ModifyTime[T, D] {
+    override def addTo(
+      temporal: T,
+      amount: D,
+    ): T = amount.addTo(temporal).asInstanceOf[T]
+    override def subtractFrom(
+      temporal: T,
+      amount: D,
+    ): T = amount.subtractFrom(temporal).asInstanceOf[T]
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala]: This isn't used. I should probably remove it.

File: core/src/test/scala/com/rallyhealth/vapors/core/interpreter/TimeFunctionsSpec.scala [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    } yield Duration.between(a, b)
+  }
+
+  implicit val arbJPeriod: Arbitrary[Period] = Arbitrary {
+    for {
+      a <- arbitrary[LocalDate]
+      b <- arbitrary[LocalDate]
+    } yield Period.between(a, b)
+  }
+  // $COVERAGE-ON$
```
Comment (jeffmay) [inline core/src/test/scala/com/rallyhealth/vapors/core/interpreter/TimeFunctionsSpec.scala:252]: Created https://jira.rallyhealth.com/browse/PRO-1039 as a follow-up

File: core/src/main/scala/com/rallyhealth/vapors/core/algebra/Expr.scala [hunk 8]
```diff
@@ -190,6 +191,15 @@ object Expr {
...
   }
 
+  final case class CustomFunction[V, A : Typeable, R, P](
+    inputExpr: Expr[V, A, P],
+    name: String,
+    evaluate: A => R,
+    capture: CaptureP[V, R, P],
+  ) extends Expr[V, R, P] { // TODO: Should this return a Try[R]? Either[Failure, R]?
+    override def visit[G[_]](v: Visitor[V, P, G]): G[R] = v.visitCustomFunction(this)
+  }
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/core/algebra/Expr.scala]: This is missing documentation.

Comment (paul-a-kennedy-rally) [reply core/src/main/scala/com/rallyhealth/vapors/core/algebra/Expr.scala]: Any hope in getting some?

File: core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala (hunk 6 above)
Comment (paul-a-kennedy-rally) [reply core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala]: Agreed.

File: core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala (hunk 3 above)
Comment (paul-a-kennedy-rally) [reply core/src/main/scala/com/rallyhealth/vapors/core/time/ModifyTime.scala:50]: Can you add this as documentation?  I like your articulation here, but this review is ephemeral...

File: core/src/test/scala/com/rallyhealth/vapors/core/example/Snippets.scala [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+import cats.Id
+import com.rallyhealth.vapors.core.algebra.Expr
+import com.rallyhealth.vapors.core.data.{FactTable, FactType}
+import com.rallyhealth.vapors.core.dsl._
+
+import java.time.temporal.ChronoUnit
+import java.time._
+
+class Snippets(val clock: Clock) {
```
Comment (paul-a-kennedy-rally) [inline core/src/test/scala/com/rallyhealth/vapors/core/example/Snippets.scala:11]: Can we have a better name than Snippets?  AgeAndDobExpressions?

**Review by paul-a-kennedy-rally — APPROVED** [review]
Most of what I'd like addressed are naming and documentation.  Given we're under a deadline, these can be fixed later.

File: core/src/main/scala/com/rallyhealth/vapors/core/algebra/Expr.scala [hunk 10]
```diff
@@ -190,6 +191,15 @@ object Expr {
     override def visit[G[_]](v: Visitor[V, P, G]): G[R] = v.visitUsingDefinitions(this)
   }
 
+  final case class CustomFunction[V, A : Typeable, R, P](
+    inputExpr: Expr[V, A, P],
+    name: String,
+    evaluate: A => R,
+    capture: CaptureP[V, R, P],
+  ) extends Expr[V, R, P] { // TODO: Should this return a Try[R]? Either[Failure, R]?
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/core/algebra/Expr.scala:210]: I created this ticket for follow-up: https://jira.rallyhealth.com/browse/PRO-1040

File: core/src/main/scala/com/rallyhealth/vapors/core/dsl/TimeFunctions.scala [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+import com.rallyhealth.vapors.core.algebra.{CaptureP, Expr}
+import com.rallyhealth.vapors.core.data.FactTable
+import com.rallyhealth.vapors.core.time.{CountTime, ModifyTime}
+import shapeless._
+
+import java.time.{Clock, Instant, LocalDate}
+
+trait TimeFunctions {
+
+  // TODO: Can this be converted to Addition[T] if we make the second argument a different type? Maybe AddLeft[T, D]?
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/core/dsl/TimeFunctions.scala:12]: Created https://jira.rallyhealth.com/browse/PRO-1041 to address this one.

File: core/src/test/scala/com/rallyhealth/vapors/core/example/Snippets.scala (hunk 9 above)
Comment (jeffmay) [reply core/src/test/scala/com/rallyhealth/vapors/core/example/Snippets.scala:11]: I was thinking of this as similar to `example.FactTypes`... Maybe `example.FactExpressions`?

**Review by darmbrus — APPROVED** [review]
